Include/internal/pycore_pythread.h

```diff
@@ -106,6 +106,36 @@ PyAPI_FUNC(PyLockStatus) PyThread_acquire_lock_timed_with_retries(
     PyThread_type_lock,
     PY_TIMEOUT_T microseconds);
 
+PyAPI_FUNC(unsigned long long) PyThread_get_thread_ident_ex(void);
+
+/* Thread joining APIs.
+ *
+ * These APIs have a strict contract:
+ *  - Either PyThread_join_thread or PyThread_detach_thread must be called
+ *    exactly once with the given handle.
+ *  - Calling neither PyThread_join_thread nor PyThread_detach_thread results
+ *    in a resource leak until the end of the process.
+ *  - Any other usage, such as calling both PyThread_join_thread and
+ *    PyThread_detach_thread, or calling them more than once (including
+ *    simultaneously), results in undefined behavior.
+ */
+PyAPI_FUNC(int) PyThread_start_joinable_thread(void (*func)(void *),
+                                               void *arg,
+                                               unsigned long long* ident,
+                                               Py_uintptr_t* handle);
+/*
+ * Join a thread started with `PyThread_start_joinable_thread`.
+ * This function cannot be interrupted. It returns 0 on success,
+ * a non-zero value on failure.
+ */
+PyAPI_FUNC(int) PyThread_join_thread(Py_uintptr_t);
+/*
+ * Detach a thread started with `PyThread_start_joinable_thread`, such
+ * that its resources are relased as soon as it exits.
+ * This function cannot be interrupted. It returns 0 on success,
+ * a non-zero value on failure.
+ */
+PyAPI_FUNC(int) PyThread_detach_thread(Py_uintptr_t);
```

Comment:
I added the typedefs, however I struggle to understand what the helpers would do or how they would work. Perhaps you mixed the typedef suggestion with the struct one?

Reply:
The point of the helpers is so we don't have to update code all over if we change the layout of underlying data (e.g. from `_Py_uintptr_t` to a struct containing that value and other info).  So I suppose it's a question of maintenance costs.

(It would also help with ABI compatibility if we were to make any of this "handle" stuff public API, though we certainly don't need to stress over that right now. 😄)

Reply:
Thanks for the typedefs, BTW.  I think we'll be glad later that we did it.